python/ray/train/base_trainer.py

```diff
@@ -700,18 +721,22 @@ def train_func(config):
             # Instantiate new Trainer in Trainable.
             trainer = trainer_cls(**config)
 
-            # Get the checkpoint from the train context, and use it to initialize
+            # Get the checkpoint from Tune, and use it to initialize
             # the restored trainer.
-            # This handles both worker-level and cluster-level restoration
-            # of the Train experiment.
             checkpoint = session.get_checkpoint()
             if checkpoint:
-                trainer.resume_from_checkpoint = checkpoint
-                # Always load the preprocessor from an available checkpoint
-                # Unless we are restoring the experiment and have explicitly
-                # passed in a new preprocessor
-                if not (restored and trainer.preprocessor):
-                    trainer.preprocessor = checkpoint.get_preprocessor()
+                # Set `_checkpoint_for_restoration` for auto-recovery fault-tolerance
+                # as well as manual restoration.
+                # `trainer.resume_from_checkpoint` is only ever set by the user.
+                trainer._checkpoint_for_restoration = checkpoint
```

Comment:
Previously, we overrode `trainer.resume_from_checkpoint`, so the 3 "restoration" paths could not be distinguished.

Now, `_checkpoint_for_restoration` only gets populated for cases 1 (auto-recovery FT) and 2 (manual Trainer.restore).